environmental/src/lib.rs

```diff
@@ -57,62 +56,69 @@ pub fn using<'a, T: 'a, R, S, F: FnOnce() -> R>(
 	// - that no other thread will use it; and
 	// - that we do not use the original mutating reference while the pointer.
 	// exists.
-	let original = global.with(|r| {
-		let mut b = r.borrow_mut();
-		let o = *b;
-		*b = protected as *mut T as *mut S;
-		o
-	});
-	let r = f();
-	global.with(|r| *r.borrow_mut() = original);
-	r
-}
-
-#[doc(hidden)]
-pub fn with<'r, R, S, T: 'r, F: FnOnce(&'r mut T) -> R>(
-	global: &'static LocalKey<RefCell<*mut S>>,
-	mutator: F,
-) -> Option<R> {
 	global.with(|r| {
-		let br = r.borrow_mut();
-		if *br != 0 as *mut S {
-			// safe because it's only non-zero when it's being called from using, which
-			// is holding on to the underlying reference (and not using it itself) safely.
-			unsafe {
-				Some(mutator(&mut *(*br as *mut S as *mut T)))
+		let original = {
+			let mut global = r.borrow_mut();
+			::std::mem::replace(&mut *global, Some(protected as _))
+		};
+
+		// even if `f` panics the original will be replaced.
+		struct ReplaceOriginal<'a, T: 'a + ?Sized> {
+			original: Option<*mut T>,
+			global: &'a RefCell<Option<*mut T>>,
+		}
+
+		impl<'a, T: 'a + ?Sized> Drop for ReplaceOriginal<'a, T> {
+			fn drop(&mut self) {
+				*self.global.borrow_mut() = self.original.take();
 			}
-		} else {
-			None
 		}
+
+		let _guard = ReplaceOriginal {
+			original,
+			global: r,
+		};
+
+		f()
 	})
 }
 
 #[doc(hidden)]
-#[macro_export]
-macro_rules! decl {
-	($name:ident : $t:ty) => {
-		thread_local! {
-			static $name: $crate::RefCell<*mut $t> = $crate::RefCell::new(0 as *mut $t);
+pub fn with<T: ?Sized, R, F: FnOnce(&mut T) -> R>(
+	global: &'static LocalKey<RefCell<Option<*mut T>>>,
```

Comment:
worth noting that `Option<*mut X>` has a zero-value optimization which means it takes the same size as a `*mut X` and the `None` simply occupies `NULL` (platform-dependent) in the pointer space.